app/controllers/incomeSources/add/CheckUKPropertyDetailsController.scala

```diff
@@ -16,60 +16,147 @@
-      andThen retrieveIncomeSources).async {
-      implicit user =>
-        handleRequest(
-          isAgent = false
-        )
-    }
+  def getHomePageUrl(isAgent: Boolean): Call = {
+    if (isAgent) controllers.routes.HomeController.showAgent else
+      controllers.routes.HomeController.show()
+  }
+
+  def getUKPropertyReportingMethodUrl(isAgent: Boolean, id: String): Call = {
+    if (isAgent) controllers.incomeSources.add.routes.UKPropertyReportingMethodController.showAgent(id) else
+      controllers.incomeSources.add.routes.UKPropertyReportingMethodController.show(id)
+  }
+
+  def getErrorHandler(isAgent: Boolean): FrontendErrorHandler with ShowInternalServerError = {
+    if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+  }
+
+  def show(): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        isAgent = false
+      )
+  }
 
   def showAgent(): Action[AnyContent] = Authenticated.async {
     implicit request =>
       implicit user =>
-        getMtdItUserWithIncomeSources(incomeSourceDetailsService).flatMap {
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
           implicit mtdItUser =>
             handleRequest(
               isAgent = true
             )
         }
   }
 
-}
+  def handleRequest(isAgent: Boolean)(implicit user: MtdItUser[_]): Future[Result] = {
+    val backUrl = getBackUrl(isAgent)
+    val postAction = getSubmitUrl(isAgent)
+    val homePageUrl = getHomePageUrl(isAgent)
+    val errorHandler = getErrorHandler(isAgent)
+
+    if (isDisabled(IncomeSources)) {
+      Future.successful(Redirect(homePageUrl))
+    } else {
+      getUKPropertyDetailsFromSession(user).toOption match {
+        case Some(checkUKPropertyViewModel: CheckUKPropertyViewModel) =>
+          Future.successful(Ok(
+            checkUKPropertyDetails(viewModel = checkUKPropertyViewModel,
+              isAgent = isAgent,
+              backUrl = backUrl,
+              postAction = postAction)))
+        case None => Logger("application").error(
+          s"[CheckUKPropertyDetailsController][handleRequest] - Error: Unable to build UK property details")
+          Future.successful(errorHandler.showInternalServerError())
+      }
+    }
+  }
+
+  def submit(): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleSubmit(isAgent = false)
+  }
+
+  def submitAgent(): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleSubmit(isAgent = true)
+        }
+  }
+
+  def handleSubmit(isAgent: Boolean)(implicit user: MtdItUser[_]): Future[Result] = {
```

Comment:
this code must be common to both UK and foreign property, no? In which case, can we add as trait perhaps?

Reply:
I don't think they can be easily combined as they deal with different view models, session keys and redirects